### PR TITLE
Add HTTPClient

### PR DESCRIPTION
**Pull request by ConnorRigby**
This adds the ability to check/download an update for devices that aren't always connected/not always available to upgrade.

### CODE REVIEW

File: lib/nerves_hub/http_fwup_steam.ex [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  end
+
+  def handle_call({:get, url}, from, s) do
+    headers = [
+      {'Content-Type', 'application/octet-stream'}
+    ]
+
+    http_opts = [timeout: :infinity, autoredirect: false]
+    opts = [stream: :self, receiver: self(), sync: false]
+    :httpc.request(:get, {String.to_charlist(url), headers}, http_opts, opts, :nerves_system_test)
```
[the remark this answers is not in the document]
Comment (ConnorRigby) [reply lib/nerves_hub/http_fwup_steam.ex]: I believe this file was copy-pasted  from the `nerves_system_test` application. There are some leftovers from it.

**Comment by ConnorRigby** [comment]
Also just realized a typo when renaming that file. :hourglass:

**Comment by fhunleth** [comment]
This PR does a lot. That makes it really hard to review, so let me try to summarize what it does.

1. Replace `:httpc` with `:hackney`. I know people have strong opinions on HTTP client libraries so I'll let others comment on this.
2. Rewrite the `fwup` download code to use `:hackney`
3. Add calls to poll for firmware updates

I don't see any documentation.

I think the best route would be to take a few steps back and get some agreement on the HTTP client library change, then submit a PR for that. Once that's merged, I'd move on to number 3, but add some documentation. The fwup download rewrite seems like cleanup - which is good, but feels like something we could tackle separately.

**Comment by ConnorRigby** [comment]
@fhunleth i didn't change the HTTP library. It was `hackney` to begin with, never `httpc` as far as i know. 

Because of renaming the file, git kind of looks weird now. but all i've done is separate "normal" HTTP requests from the "Fwup streaming" ones. 

I think it only looks like a lot because of the file rename weirdness.

**Comment by ConnorRigby** [comment]
The contents of `lib/nerves_hub/http_fwup_stream.ex` remain the same as `lilb/nerves_hub/http_client.ex`, the name of the file and module just needed to be changed.

File: lib/nerves_hub/http_fwup_stream.ex [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+  def handle_call({:get, url}, from, s) do
+    headers = [
+      {'Content-Type', 'application/octet-stream'}
+    ]
+
+    http_opts = [timeout: :infinity, autoredirect: false]
+    opts = [stream: :self, receiver: self(), sync: false]
+
+    :httpc.request(
```
Comment (mobileoverlord) [inline lib/nerves_hub/http_fwup_stream.ex]: If were going to be adding hackney, then it doesn't make sense to continue to use :httpc here

Comment (mobileoverlord) [reply lib/nerves_hub/http_fwup_stream.ex]: I guess we could tackle that in a follow up.

**Pull request merged**